ReadMe.md
# Vault

## Abstract

Provides a PowerShell command-line interface to the "Web Credentials" section of the [Windows Password Vault](https://docs.microsoft.com/en-us/uwp/api/Windows.Security.Credentials.PasswordVault) (called "Credential Manager" in the Windows GUI). This allows for a decent measure of protection for credentials at rest. As always, the best defense for secrets is to make sure only authorized entities can access them, and that authentication of authorized parties is as strong as possible.

## Quick Start

### Example 1: Get a Credential from the Windows Credential Manager

You need to get a stored credential from the Windows Credential Manager. You may need to view the password in clear-text.

```PowerShell
# get the credential without the password by the username
Get-VaultEntry -UserName test


UserName Resource         Password Properties
-------- --------         -------- ----------
test     http://test.com/          {[hidden, False], [applicationid, 00000000-0000-0000-0000-000000000000], [application, ]}

# get the same credential, this time returning the clear-text password

Get-VaultEntry -UserName test -IncludePassword


UserName Resource         Password Properties
-------- --------         -------- ----------
test     http://test.com/ test     {[hidden, False], [applicationid, 00000000-0000-0000-0000-000000000000], [application, ]}

```

### Example 2: Store a Credential in the Windows Credential Manager

You need to use an API key in clear-text with an API, but you do not want to store the key in clear-text.
You want the key to be decrypted by the same user on the same computer it was encrypted on.

```PowerShell
# get the credential
$credToAdd = Get-Credential

# add it to the Password Vault
Add-VaultEntry -Credential $credToAdd -Resource http://test.com

```

### Example 3: Remove a Credential from the Windows Credential Manager

You need to remove a credential from the Windows Credential Manager.

```PowerShell
# remove the credential by resource
Remove-VaultEntry -Resource http://test.com/ -Verbose
```


ReleaseNotes.md
# 2019/07/23 'Indignant Hurricane of Annihilation'

Pulled Windows Credential Manager code from an older project named "OMTCredential". Cleaned up code a bit and fixed documentation.


Vault/public/Add-VaultEntry.ps1
<#
.SYNOPSIS
Adds a credential to the Windows Credential Manager.

.PARAMETER PsCredential
A [pscredential] that represents the username and password you want to store in the vault.

[pscredential] objects can be made with the `Get-Credential` cmdlet.

.PARAMETER Resource
The name to store the credential under in the vault.

.EXAMPLE

$credentialToAdd = Get-Credential

Add-VaultEntry -PsCredential $credentialToAdd -Resource ApiUser

.NOTES

#>
function Add-VaultEntry {

    [CmdletBinding()]

    param (

        [Parameter(Mandatory, ValueFromPipelineByPropertyName, ValueFromPipeline)]
        [ValidateNotNullOrEmpty()]
        [pscredential]
        $PsCredential,

        [Parameter(Mandatory, ValueFromPipelineByPropertyName)]
        [ValidateNotNullOrEmpty()]
        [uri]
        $Resource

    )

    begin {

        $ErrorActionPreference = 'Stop'

        $PasswordVault = InitializePasswordVault

    }

    process {

        Write-Verbose "[$($MyInvocation.MyCommand.Name)]: $($Resource): Adding credential for user '$($PsCredential.UserName)'..."

        try {

            $splat = @{

                TypeName     = 'Windows.Security.Credentials.PasswordCredential'

                ArgumentList = $Resource, $PsCredential.UserName, (ConvertSecureStringToString -SecureString $PsCredential.Password)

            }

            $vaultEntry = New-Object @splat

            $PasswordVault.Add($vaultEntry)

        }

        catch {

            Write-Error -ErrorAction Stop -Exception $_.Exception

        }

        Write-Verbose "[$($MyInvocation.MyCommand.Name)]: $($Resource): Added credential for user '$($PsCredential.UserName)'."

    }

    end {

    }

}

Vault/public/Remove-VaultEntry.ps1
<#
.SYNOPSIS
Removes credentials from the Windows Password Vault.

.PARAMETER Resource
Removes all credentials where the resource (which is a URI) matches this value. Case-sensitive.

.PARAMETER UserName
Removes all credentials where the username matches this value. Case-sensitive.

.PARAMETER RemoveAll
Removes all credentials for the current user.

.PARAMETER Force
WARNING: Crendential entries removed can not be recovered.
Bypasses the confirmation prompt removing entries.

.EXAMPLE
Remove-VaultEntry -UserName test

UserName Resource         Password Properties
-------- --------         -------- ----------
test     http://test.com/          {[hidden, False], [applicationid, 00000000-0000-0000-0000-000000000000], [application, ]}

Gets all credentials with the username "test".

.EXAMPLE
Remove-VaultEntry -Resource http://test.com/

UserName Resource         Password Properties
-------- --------         -------- ----------
test     http://test.com/          {[hidden, False], [applicationid, 00000000-0000-0000-0000-000000000000], [application, ]}

Gets all credentials with the resource "http://test.com/". Note that URIs in this case are case-sensitive--http://test.com/  is not the same as http://tEsT.cOm/ as far as this cmdlet is concerned.

.NOTES

The underlying class, [Windows.Security.Credentials.PasswordVault], is oddly case-sensitve.

#>
function Remove-VaultEntry {

    [CmdletBinding(DefaultParameterSetName = 'FindAllByResource', ConfirmImpact = "high", SupportsShouldProcess)]

    param (

        [Parameter(Mandatory, ValueFromPipelineByPropertyName, ValueFromPipeline, ValueFromRemainingArguments, ParameterSetName = 'FindAllByResource')]
        [ValidateNotNullOrEmpty()]
        [uri]
        $Resource,

        [Parameter(Mandatory, ValueFromPipelineByPropertyName, ValueFromPipeline, ValueFromRemainingArguments, ParameterSetName = 'FindAllByUserName')]
        [ValidateNotNullOrEmpty()]
        [string]
        $UserName,

        [Parameter(ValueFromPipelineByPropertyName, ValueFromPipeline, ParameterSetName = 'RetrieveAll')]
        [switch]
        $All,

        [switch]
        $Force

    )

    begin {

        $ErrorActionPreference = 'Stop'

        $PasswordVault = InitializePasswordVault

    }

    process {

        if ($PSCmdlet.ShouldProcess("Removing credentials--they will not be recoverable")) {

            try {

                $value = if ($Resource) { Write-Output $Resource } else { Write-Output $UserName }

                $windowsSecurityCredential = GetPasswordVaultEntry -ErrorAction Stop -Mode $PsCmdlet.ParameterSetName -Value $value

                $int_ShouldContinueCountValue = ($windowsSecurityCredential | Measure-Object).Count

                $shouldContinueCaption = "If you continue, $($int_ShouldContinueCountValue) credential(s) will be deleted. There is no way to recover them."
                $shouldContinueMessage = "Please make sure you have taken the necessary precautions. Continue?"

                if ($Force -or $PsCmdlet.ShouldContinue($shouldContinueMessage, $shouldContinueCaption)) {

                    $PasswordVault.Remove($windowsSecurityCredential)

                }

            }

            catch {

                Write-Error -ErrorAction Stop -Exception $_.Exception

            }

        }
    }

    end {

    }

}

Vault/private/ConvertWindowsSecurityCredentialToPSCredential.ps1
function ConvertWindowsSecurityCredentialToPSCredential {

    [CmdletBinding()]

    [Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSAvoidUsingPlainTextForPassword', '')]

    param (

        [Parameter(Mandatory, ValueFromPipelineByPropertyName, ValueFromPipeline)]
        [ValidateNotNullOrEmpty()]
        [Windows.Security.Credentials.PasswordCredential]
        $WindowsSecurityCredential

    )

    begin {

    }

    process {

        try {

            Write-Verbose "[$($MyInvocation.MyCommand.Name)]: Converting input to pscredential type."

            $psCredential = New-Object -TypeName pscredential -ArgumentList $WindowsSecurityCredential.UserName, ($WindowsSecurityCredential.Password | ConvertStringToSecureString)

            Add-Member -ErrorAction Stop -MemberType NoteProperty -Value $WindowsSecurityCredential.Resource -Name Resource -InputObject $psCredential

            Write-Output $psCredential

        } catch {

            Write-Error -ErrorAction Stop -Exception $_.Exception

        }

    }

    end {

    }

}

Vault/public/Get-VaultEntry.ps1
<#
.SYNOPSIS
Gets credentials from the Windows Password Vault.

.PARAMETER Resource
Gets credentials where the resource matches this value. Case-sensitive.

.PARAMETER UserName
Gets credentials where the username matches this value. Case-sensitive.

.PARAMETER All
Gets all credentials for the current user.

.PARAMETER IncludePassword
WARNING: Password will be written in clear-text.

Includes passwords in returns.

.PARAMETER AsPsCredential
Returns a [pscredential] for each credential entry. The password is always included when this mode is selected.

.PARAMETER Force
Bypasses the confirmation prompt when a credential would include clear-text passwords.

.EXAMPLE
Get-VaultEntry -UserName test

UserName Resource         Password Properties
-------- --------         -------- ----------
test     http://test.com/          {[hidden, False], [applicationid, 00000000-0000-0000-0000-000000000000], [application, ]}

Gets all credentials with the username "test".

.EXAMPLE
Get-VaultEntry -Resource http://test.com/

UserName Resource         Password Properties
-------- --------         -------- ----------
test     http://test.com/          {[hidden, False], [applicationid, 00000000-0000-0000-0000-000000000000], [application, ]}

Gets all credentials with the resource "http://test.com/".
Note: URIs are case-sensitive--http://test.com/  is not the same as http://tEsT.cOm/.

.NOTES

The underlying class, [Windows.Security.Credentials.PasswordVault], is oddly case-sensitve.

#>
function Get-VaultEntry {

    [CmdletBinding(DefaultParameterSetName = 'RetrieveAll', ConfirmImpact = "high")]

    param (

        [Parameter(Mandatory, ValueFromPipelineByPropertyName, ValueFromPipeline, ParameterSetName = 'FindAllByResource')]
        [ValidateNotNullOrEmpty()]
        [uri]
        $Resource,

        [Parameter(Mandatory, ValueFromPipelineByPropertyName, ParameterSetName = 'FindAllByUserName')]
        [ValidateNotNullOrEmpty()]
        [string]
        $UserName,

        [Parameter(ValueFromPipelineByPropertyName, ParameterSetName = 'RetrieveAll')]
        [switch]
        $All,

        [Parameter(ValueFromPipelineByPropertyName, ParameterSetName = 'IncludePassword')]
        [Parameter(ValueFromPipelineByPropertyName, ParameterSetName = 'FindAllByResource')]
        [Parameter(ValueFromPipelineByPropertyName, ParameterSetName = 'FindAllByUserName')]
        [Parameter(ValueFromPipelineByPropertyName, ParameterSetName = 'RetrieveAll')]
        [switch]
        $IncludePassword,

        [Parameter(ValueFromPipelineByPropertyName, ParameterSetName = 'AsPsCredential')]
        [Parameter(ValueFromPipelineByPropertyName, ParameterSetName = 'FindAllByResource')]
        [Parameter(ValueFromPipelineByPropertyName, ParameterSetName = 'FindAllByUserName')]
        [Parameter(ValueFromPipelineByPropertyName, ParameterSetName = 'RetrieveAll')]
        [switch]
        $AsPsCredential,

        [Parameter(ValueFromPipelineByPropertyName, ParameterSetName = 'IncludePassword')]
        [Parameter(ValueFromPipelineByPropertyName, ParameterSetName = 'FindAllByResource')]
        [Parameter(ValueFromPipelineByPropertyName, ParameterSetName = 'FindAllByUserName')]
        [Parameter(ValueFromPipelineByPropertyName, ParameterSetName = 'RetrieveAll')]
        [switch]
        $Force

    )

    begin {

        $ErrorActionPreference = 'Stop'

    }

    process {

        Write-Verbose "[$($MyInvocation.MyCommand.Name)]: Password-inclusion Mode is set to $($($IncludePassword, $AsPsCredential) -contains $true)."

        try {

            $value = if ($Resource) { Write-Output $Resource } else { Write-Output $UserName }

            $result = GetPasswordVaultEntry -Mode $PsCmdlet.ParameterSetName -Value $value

            if ($IncludePassword.IsPresent) {

                $shouldContinueMessage = "Please make sure you have taken the necessary precautions."

                $shouldContinueCaption = "If you continue, credential passwords will be decrypted and may be written to the host in clear-text. Continue?"

                if ($Force -or $PsCmdlet.ShouldContinue($shouldContinueMessage, $shouldContinueCaption)) {

                    $result.RetrievePassword()

                    Write-Output $result

                }

            }

            if ($AsPsCredential.IsPresent) {

                Write-Verbose "[$($MyInvocation.MyCommand.Name)]: Starting pscredential conversion."

                $result.RetrievePassword()

                Write-Output ($result | ConvertWindowsSecurityCredentialToPSCredential)

                Write-Verbose "[$($MyInvocation.MyCommand.Name)]: Done with pscredential conversion."

            }

            if ($AsPsCredential.IsPresent -eq $false -and $IncludePassword.IsPresent -eq $false) {

                Write-Output $result

            }

        } catch {

            Write-Error -ErrorAction Stop -Exception $_.Exception

        }

    }

    end {

    }

}